tests/test_agents.py

```diff
@@ -709,3 +711,47 @@ def test_agent_prompt_collection_validations(
     else:
         with pytest.raises(result):
             AgentSettings(**kwargs)
+
+
+@pytest.mark.asyncio
+async def test_deepcopy_env(agent_test_settings: Settings) -> None:
+    await get_directory_index(settings=agent_test_settings)  # Trigger build
+
+    question = "How can you use XAI for chemical property prediction?"
+    env = GradablePaperQAEnvironment(
+        query=QueryRequest(query=question, settings=agent_test_settings),
+        docs=Docs(),
+    )
+
+    # 1. Rollout until after gather evidence
+    await env.reset()
+    for tool_call in (
+        ToolCall.from_name(
+            "paper_search",
+            query="XAI for chemical property prediction",
+            min_year=2018,
+            max_year=2024,
+        ),
+        ToolCall.from_name("gather_evidence", question=question),
+    ):
+        await env.step(ToolRequestMessage(tool_calls=[tool_call]))
+
+    # 2. Now we deepcopy the environment
+    env_copy = deepcopy(env)
+    assert env.state == env_copy.state
+
+    # 3. Generate an answer for both, and confirm they are identical
+    gen_answer_action = ToolRequestMessage(
+        tool_calls=[ToolCall.from_name("gen_answer", question=question)]
+    )
+    _, _, done, _ = await env.step(gen_answer_action)
+    assert done
+    assert not env.state.answer.could_not_answer
+    assert env.state.answer.used_contexts
+    _, _, done, _ = await env_copy.step(gen_answer_action)
```

Comment:
Is there a chance this could be flaky given it’s inference based?

Reply:
Yeah we definitely can be flaky, I added `pytest.mark.flaky` just now